tests/test_requests.py

```diff
@@ -315,6 +315,26 @@ def test_headers_on_session_with_None_are_not_sent(self, httpbin):
         prep = ses.prepare_request(req)
         assert 'Accept-Encoding' not in prep.headers
 
+    def test_headers_preserve_order(self, httpbin):
+        """Preserve order when headers provided as OrderedDict."""
+        ses = requests.Session()
+        ses.headers = OrderedDict()
```

Comment:
All tests pass! I may cut a new release tonight, and would love to include this, so would love an answer to this question@piotrjurkiewicz :)